File: C/Leetcode/AddTwoNumbers.c

#include <stdio.h>
#include <stdlib.h>

int max(int a, int b) {
    return (a > b) ? a : b;
}

void reverse(int arr[], int len) {
    for (int i = 0; i < len / 2; ++i) {
        int temp = arr[i];
        arr[i] = arr[len - i - 1];
        arr[len - i - 1] = temp;
    }
}

int main() {
    int a[] = {9, 9, 9};
    int b[] = {1};
    int lenA = sizeof(a) / sizeof(a[0]);
    int lenB = sizeof(b) / sizeof(b[0]);

    reverse(a, lenA);
    reverse(b, lenB);

    int maxLen = max(lenA, lenB);
    int* result = (int*)malloc((maxLen + 1) * sizeof(int)); // +1 for possible carry
    int carry = 0, sum, i;

    for (i = 0; i < maxLen; i++) {
        int digitA = (i < lenA) ? a[i] : 0;
        int digitB = (i < lenB) ? b[i] : 0;

        sum = digitA + digitB + carry;
        result[i] = sum % 10;
        carry = sum / 10;
    }

    if (carry) {
        result[i++] = carry;
    }

    // Print in reverse to show the correct number
    printf("Result: ");
    for (int j = i - 1; j >= 0; j--) {
        printf("%d", result[j]);
    }
    printf("\n");

    free(result);
    return 0;
}

File: JavaScript/Leetcode/AddTwoNumbers.js

let a = [9, 9, 9];
let b = [1];
let c = [];

// Reverse the arrays to start from the least significant digit
a.reverse();
b.reverse();

let carry = 0;
let maxLen = Math.max(a.length, b.length);

for (let i = 0; i < maxLen; i++) {
    let digitA = i < a.length ? a[i] : 0;
    let digitB = i < b.length ? b[i] : 0;

    let sum = digitA + digitB + carry;
    c.push(sum % 10);
    carry = Math.floor(sum / 10);
}

if (carry > 0) {
    c.push(carry);
}

// Reverse the result back
c.reverse();



